[Apple Pay] add a `features` that allows for direct detection of Apple Pay features (similar to the more general `version`)

dcrousso opened the pull request:
#### 0f50dd00e346e4e6bbcd5c491763a0e6dd147cbc
<pre>
[Apple Pay] add a `features` that allows for direct detection of Apple Pay features (similar to the more general `version`)
<a href="https://bugs.webkit.org/show_bug.cgi?id=243269">https://bugs.webkit.org/show_bug.cgi?id=243269</a>
&lt;rdar://problem/97529973&gt;

Reviewed by Aditya Keerthi.

We currently offer the `version` as an indirect way for developers to answer the question &quot;does the
current device support this feature&quot; since the developer can look up the version that supports that
feature &lt;<a href="https://developer.apple.com/documentation/apple_pay_on_the_web/apple_pay_on_the_web_version_history">https://developer.apple.com/documentation/apple_pay_on_the_web/apple_pay_on_the_web_version_history</a>&gt;.

While there&apos;s nothing wrong with this, it&apos;s a pretty inflexible (monotonically increasing integer)
and not directly clear approach (requires looking at documentation to know what something means).

We can instead have a list of `features` that contains an `enum`, allowing developers to call out
desired features by name, allowing for more more options and avoiding looking at documentation.

* Source/WebCore/Modules/applepay/ApplePayFeature.idl: Added.
* Source/WebCore/Modules/applepay/ApplePayFeature.h: Added.
Add enum values for each of the features added in the last few years.

* Source/WebCore/Modules/applepay/ApplePayRequestBase.idl:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.h:
Allow the `ApplePayFeature` to be provided when creating the `PaymentRequest`.

* Source/WebCore/Modules/applepay/cocoa/PaymentAPIVersionCocoa.mm:
(WebCore::PaymentAPIVersion::current):
Bump the current Apple Pay version (hopefully for the last time, as we should use `ApplePayFeature`
for features going forward).

* LayoutTests/http/tests/paymentrequest/paymentrequest-features.https.html: Added.
* LayoutTests/http/tests/paymentrequest/paymentrequest-features.https-expected.txt: Added.

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-additionalShippingMethods.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-additionalShippingMethods.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-automaticReloadPaymentRequest.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-automaticReloadPaymentRequest.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-multiTokenContexts.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-multiTokenContexts.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-recurringPaymentRequest.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-recurringPaymentRequest.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayPaymentCompleteDetails-orderDetails.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayPaymentCompleteDetails-orderDetails.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/paymentrequest-couponCode.https.html:
* LayoutTests/http/tests/paymentrequest/paymentrequest-couponCode.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/paymentrequest-shippingContactEditingMode.https.html:
* LayoutTests/http/tests/paymentrequest/paymentrequest-shippingContactEditingMode.https-expected.txt:
Add tests for the relevant `ApplePayFeature`.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/253276@main">https://commits.webkit.org/253276@main</a>
</pre>

pxlcoder commented on LayoutTests/http/tests/paymentrequest/paymentrequest-features.https.html:
```diff
@@ -0,0 +1,80 @@
+<body>
+<script>
+
+description("Tests for providing `features` as part pf `ApplePayRequest`.");
```
Typo: `pf`.

pxlcoder commented on LayoutTests/http/tests/paymentrequest/paymentrequest-features.https.html:
```diff
@@ -0,0 +1,80 @@
+
+    var supportsTouchEvents = "TouchEvent" in window;
+    if (supportsTouchEvents && window.testRunner && testRunner.runUIScript) {
+        testRunner.runUIScript(`(function() { uiController.singleTapAtPoint(${x}, ${y}, function() { }); })();`, function () { });
```
I think you can use `UIHelper.tapElement` here.

pxlcoder commented on LayoutTests/http/tests/paymentrequest/paymentrequest-features.https.html:
```diff
@@ -0,0 +1,80 @@
+<head>
+<meta charset="utf-8">
+<script src="/js-test-resources/ui-helper.js"></script>
+<script src="/resources/js-test-pre.js"></script>
```
Nit: New tests don't need `js-test-pre.js` and `js-test-post.js`, can just use `js-test.js`.

pxlcoder commented on Source/WebCore/Modules/applepay/cocoa/PaymentAPIVersionCocoa.mm:
```diff
@@ -34,27 +34,7 @@
-#endif
-        return 7;
-    }();
-    return current;
```
Is is safe to remove these values? Aren't some of these features dependent on OS version?

dcrousso replied:
yes because `features` is not guarded and therefore will be available in all future builds regardless of OS version

also, we only throw an error if the given `version` is greater than this value, so the only way this would be problematic (i.e. a compat problem) is if some page out on the web was already using `version: 15` as that would no longer throw an error

pxlcoder replied:
(we talked about this offline, and need to preserve these values since we ship latest WebKit to older versions of macOS)

dcrousso committed:
[Apple Pay] add a `features` that allows for direct detection of Apple Pay features (similar to the more general `version`)
https://bugs.webkit.org/show_bug.cgi?id=243269
<rdar://problem/97529973>

Reviewed by Aditya Keerthi.

We currently offer the `version` as an indirect way for developers to answer the question "does the
current device support this feature" since the developer can look up the version that supports that
feature <https://developer.apple.com/documentation/apple_pay_on_the_web/apple_pay_on_the_web_version_history>.

While there's nothing wrong with this, it's a pretty inflexible (monotonically increasing integer)
and not directly clear approach (requires looking at documentation to know what something means).

We can instead have a list of `features` that contains an `enum`, allowing developers to call out
desired features by name, allowing for more more options and avoiding looking at documentation.

* Source/WebCore/Modules/applepay/ApplePayFeature.idl: Added.
* Source/WebCore/Modules/applepay/ApplePayFeature.h: Added.
Add enum values for each of the features added in the last few years.

* Source/WebCore/Modules/applepay/ApplePayRequestBase.idl:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.h:
Allow the `ApplePayFeature` to be provided when creating the `PaymentRequest`.

* Source/WebCore/Modules/applepay/cocoa/PaymentAPIVersionCocoa.mm:
(WebCore::PaymentAPIVersion::current):
Bump the current Apple Pay version (hopefully for the last time, as we should use `ApplePayFeature`
for features going forward).

* LayoutTests/http/tests/paymentrequest/paymentrequest-features.https.html: Added.
* LayoutTests/http/tests/paymentrequest/paymentrequest-features.https-expected.txt: Added.

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-additionalShippingMethods.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-additionalShippingMethods.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-automaticReloadPaymentRequest.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-automaticReloadPaymentRequest.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-multiTokenContexts.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-multiTokenContexts.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-recurringPaymentRequest.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-recurringPaymentRequest.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayPaymentCompleteDetails-orderDetails.https.html:
* LayoutTests/http/tests/paymentrequest/ApplePayPaymentCompleteDetails-orderDetails.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/paymentrequest-couponCode.https.html:
* LayoutTests/http/tests/paymentrequest/paymentrequest-couponCode.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/paymentrequest-shippingContactEditingMode.https.html:
* LayoutTests/http/tests/paymentrequest/paymentrequest-shippingContactEditingMode.https-expected.txt:
Add tests for the relevant `ApplePayFeature`.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/253276@main
